proposals/p1360.md

```diff
@@ -0,0 +1,90 @@
+# Change raw string literal syntax: [#]\*""" always starts a block string
+
+<!--
+Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+Exceptions. See /LICENSE for license information.
+SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+-->
+
+[Pull request](https://github.com/carbon-language/carbon-lang/pull/1360)
+
+<!-- toc -->
+
+## Table of contents
+
+-   [Problem](#problem)
+-   [Background](#background)
+-   [Proposal](#proposal)
+-   [Details](#details)
+-   [Rationale](#rationale)
+-   [Alternatives considered](#alternatives-considered)
+
+<!-- tocstop -->
+
+## Problem
+
+Under current design of string literals, users may make assumptions that a
+starting `[#]*"""` represents a block string and misunderstand the syntax.
+
+## Background
+
+The design of
+[string literals](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/lexical_conventions/string_literals.md)
+specifies a block string literal to start with `[#]*"""`. Users may take for
+granted the other way around, where any string literal starting with `[#]*"""`
+is a block string literal. This does not hold true, however. Two counter-cases
+are `"""abc"""` represents three tokens `""`, `"abc"` and `""`, and `#"""#`
+which is equivalent to `"\""`. Neither is a block string literal and may be
+visually confusing.
+
+## Proposal
+
+Always interpret `[#]*"""` as the start of a block string literal, even if the
+result is invalid code.
+
+## Details
+
+Two changes are made in this proposal. First, `"""` always starts a simple block
+string literal. Instead of the three tokens described above, `"""abc"""` will be
+parsed as an invalid simple block string literal due to quotes in the file type
+`abc"""`. Users can use extra spaces like `"" "abc" ""` to represent the three
+tokens represented in the original design. Second, `[#]+"""` always starts a raw
+block string literal. `#"""abc"""#` becomes an invalid block string literal,
+compared to a valid single-line string literal in the original design. More
+details can be found
+[here](/docs/design/lexical_conventions/string_literals.md).
+
+## Rationale
+
+This principle helps make Carbon code
+[easy to read, understand, and write](/docs/project/goals.md#code-that-is-easy-to-read-understand-and-write),
+because it avoids confusion on the type of certain string literals.
+
+## Alternatives considered
```

Comment:
I think there are other alternatives we should at least document here...

We could allow a non-escape based disambiguation that focuses on making the opening marker not ambiguous between single-line and block. I can imagine two ways of doing this.

-   a) We could allow some non-quote marker after the open quote similar to C++. Either `#"(")"#` or `#"#"#"#` come to mind.

-   b) We could use different quotes in various ways that allow: `#'"'#`.
    -    b1) `/#+'/` could be an additional way to start a single line string literal, without removing anything
    -    b2) `/#+'/` could always start a single line string literal, and `/#+"` could always start a block string literal

These all seem to have problems though that should be covered in the alternatives considered.

(a) IMO hurts readability and makes it harder to adddress a complaint of a collision by adding a(onther) set of `#`s around the contents. The first time, you'd have to modify things more than that. It also loses the visual connection of the `"`s surrounding the quoted-text.

(b) Adds a lot of complexity no matter how we slice it, and doesn't fully remove a lurking surprise in the lexical structure.

Fundamentally, (a), (b), (c) this proposal, and (d) the status quo are moving around *where* there is a surprise, but not eliminating it.

The source of the surprise is that the introducing syntax for single line string literals in a *prefix* of multiline string literals, including in the raw case. The result is that we don't know which we're seeing until its too late to avoid surprises.

But there is yet another alternative, suggested @zygoloid after I gave a long series of bad suggestions, let's call it (e).

In (e), we change block string literals to always use `'''`, never `"""`. Everything else stays the same. If/when we have character literals of any form, we reject `''`, when we see `''` we know the next character must be `'` and it is a block string literal. Raw block string literals are `#+'''`. Etc. Python already allows `'''` as well, although it is rarely used to my knowledge.

With (e), `#"""#` is the same as `"\""` and doesn't require anything to recognize. You can even write `#""""A python string here!""""#`.

For teaching, this even seems nice -- its a "triply quoted" string instead of a "triple-double quoted" string. ;]

WDYT? Both @zygoloid and I quite like this, but we should check with @KateGregory as well.

Reply:
Just FYI, seems all leads are happy here and haven't heard huge concerns. It is a bit inventive, but I think folks are roughly comfortable so let's consider the path forward.

Reply:
Per discussion, I believe the intent is to also disallow `"""` (triple double quote), rather than interpreting it as two string literals. This would help reduce the confusion with Python (where `'''` is _supported_, but `"""` is _recommended_ under PEP 257).

Of course, `#"""` is still an issue -- but I think it sounds like the leads want to support that use-case?

Reply:
Just to note, I think it'd be good to go through the discussed options and some advantages and disadvantages of each in the alternatives. If you're trying to find a good example of how to do this, something like the alternatives from one of these:

Explicitly calling out advantages and disadvantages:
https://github.com/carbon-language/carbon-lang/blob/trunk/proposals/p0257.md#alternatives-considered

Using prose to talk about options, but a separate section for each significant option:
https://github.com/carbon-language/carbon-lang/blob/trunk/proposals/p0722.md#alternatives-considered

Reply:
Added discussed options to the alternatives.